### DISCUSSION TITLE
TensorMechanics force boundary condition

### DISCUSSION THREAD

**Discussion by ajdubas** [idaholab/moose#16104] [Q&A Modules: General] [answered]
Hi there,

I am working on a TensorMechanics problem using MOOSE and am trying to implement a force boundary condition.  To check I am implementing it correctly I am using a simple beam theory set-up and looking through the example I have tried both a NodalKernels/ConstantRate and NodalKernels/UserForcingFunctionNodalKernal with a transient ramp to the desired load.  Both of these results in a displacement two orders of magnitude greater than expected (they are consistent with each other in their answer), so am I using the correct object?  and secondly am I using it correctly?

Many Thanks in Advance,

Aleks

**Comment by crswong888** [comment]
`UserForcingFunctionNodalKernel` is definitely the way to go for concentrated beam forces, since its more flexible than `ConstantRate`. But for a static load, either should do.

I will need to see your input file to understand your issue and then I am sure we can resolve it swiftly. Please provide a link to your input file or attach it here by dragging it into the comment box (you might have to zip it if GitHub says "We don't support that file type"), or just copy and paste the code here.

**Comment by ajdubas** [comment]
Please find input file attached.

Note this uses small strain formulation as that is what the analytical validation case is based on.  The expected displacement is 0.01m at the forced end.

[structural_transient_force_3D.i.zip](https://github.com/idaholab/moose/files/5510978/structural_transient_force_3D.i.zip)

**Comment by crswong888** [comment]
@ajdubas what is it exactly you are trying to model here? A cantilevered steel beam with a concentrated transverse force of -10 kN at the free-end, yes?

If so, what are you trying to accomplish by applying 10 N in the axial direction here?

```
[Functions]
  [force]
    type = PiecewiseLinear
    x = '0.0 10.0'
    y = '0.0 -10000.0'
  []
[]
```

I was thinking you were working with beam elements, so that's why I said `UserForcingFunctionNodalKernel` is the way to go. But now I see that you are using solid elements, and so this is not the case. However, I think I still have a solution for you, but first I need to be clear about what you're trying to model here.

**Reply by crswong888** [reply]
To be clear, I think you are trying to model this thing:

![Selection_001](https://user-images.githubusercontent.com/41302548/98739729-212af880-2367-11eb-8d2a-f58f18bfa418.png)

If so, do you agree that the analytical max deflection for weak-axis bending should be

![CodeCogsEqn (10)](https://user-images.githubusercontent.com/41302548/98740880-fb065800-2368-11eb-9d71-8499a4e98f5a.gif)
 
or, in accordance with Timoshenko theory:

![CodeCogsEqn (9)](https://user-images.githubusercontent.com/41302548/98740905-03f72980-2369-11eb-8e95-dd273f89593b.gif)

**Comment by ajdubas** [comment]
Yes, that is the solution I am trying to replicate - or more generally work out how to accurately apply a force boundary condition to solid elements.

My understanding of the PiecewiseLinear function ( https://mooseframework.inl.gov/source/functions/PiecewiseLinear.html ) is that the x array corresponds to time, therefore the force should ramp from 0.0N at 0.0s to -10.0kN at 10.0s if I have specified this function correctly.

**Comment by crswong888** [comment, marked as answer]
@ajdubas That's right, my mistake. By default, the ordinates for `PiecewiseLinear` correspond to time (for some reason I thought there was a `t` parameter, but all of these details are worked out with the `axis` parameter).

The reason for the large displacements is a vastly over-sized force. Keep in mind that `UserForcingFunctionNodalKernel` will apply the full magnitude of the specified forcing function to _every_ single node on the specified boundary. Therefore, the total magnitude of the force you are applying is -10000 N / node * 340 nodes = 3.4e6 N. A force that bears more semblance to the concentrated transverse beam force you want would be roughly -29.4118 N / node. So just use this number instead of -10000 and you'll be good.

Note that simply dividing the total force by the number of nodes and applying exactly this value to every single node isn't quite how we think of concentrated forces on beams. Since each node has a different _tributary area_, as we like to call it in Civil Engineering, then the force applied at each node should be proportional to this, if it were truly uniformly distributed over the cross-section (which is the assumption of beam theory). I have a very rough `MooseObject` to handle this calculation if you're interested. Just copy my files, [`PointForcingFunction3DEquivalent.C`](https://github.com/crswong888/scorpion/blob/master/src/nodalkernels/PointForcingFunction3DEquivalent.C) and [`PointForcingFunction3DEquivalent.h`](https://github.com/crswong888/scorpion/blob/master/include/nodalkernels/PointForcingFunction3DEquivalent.h) to your MOOSE app and recompile.

I've made some edits to your input file here: https://github.com/crswong888/scorpion/blob/master/inputs/users/structural_transient_force_3D.i

It currently uses my object for point forces in 3D, so you will need those files. However, I left your old blocks commented out and also updated the force magnitude to the value you'll want in those. Lastly, I changed your executioner to something that performs at least twice as well. Enjoy :)

**Reply by ajdubas** [reply]
Thank you, this works brilliantly.

Is there any way of doing this natively (i.e. just using the TensorMechanics and Contacts module without any extra code)?  It seems to me like something that should come built in, so might it be worth adding the above NodalKernel into the modules?

**Reply by crswong888** [reply]
I agree that this capability would be a useful tool to have in TensorMechanics. However, formal development of a new `MooseObject` takes time. I also don't like how my current code requires setting up multiple objects to make it work (i.e., the area postprocessor and the auxiliary kernel to store that pp value). It seems to me like this object should be able to handle the tributary areas on its own (e.g., by interpolating the shape functions to the nodes/qps with that concentrated force value - sort of like how traction loads are handled). So before I ever attempted to develop this in MOOSE, I would want to try to figure this out.

If you have the time, feel free to take my code and develop it in MOOSE yourself. See [Framework Development](https://mooseframework.inl.gov/framework_development/index.html) to learn how to do this. Once you create a PR, tag me in it, and I can help you develop it.
